### ISSUE TITLE
The role of projector layer

### ISSUE THREAD

**Issue by ZX-lang** [DefangChen/SimKD#4]
When the feature dimensions from the student and teacher are equal, such as resn32x4 and res8x4, it doesn't need projector, we try to remove the projector. But we get a terrible result. So the reason why this method works is because of the existence of the projector？

**Comment by DefangChen** [comment]
Generally, feature distributions of the teacher and student model are statistically different and cannot be directly compared in practice even their dimensions are equal. We thus follow the setting of previous KD methods (e.g., FitNet, CRD, SRRL, SemCKD) and retain this design.

Experiments show that without reusing the teacher classifier, the modified student model (containing extra projector layer) cannot achieve good results by training with regular cross-entropy loss or KD loss. This indicates the projector layer itself do not sufficiently account for the SimKD performance.

**Comment by ZX-lang** [comment]
> Generally, feature distributions of the teacher and student model are statistically different and cannot be directly compared in practice even their dimensions are equal. We thus follow the setting of previous KD methods (e.g., FitNet, CRD, SRRL, SemCKD) and retain this design.
> 
> Experiments show that without reusing the teacher classifier, the modified student model (containing extra projector layer) cannot achieve good results by training with regular cross-entropy loss or KD loss. This indicates the projector layer itself do not sufficiently account for the SimKD performance.

Thank you for your reply. So how to find a method to remove the projector is still a difficult problem.

**Comment by DefangChen** [comment]
Exactly. I believe it is of great importance to the KD research.

**Comment by ZX-lang** [comment]
> Exactly. I believe it is of great importance to the KD research.

Thank you. I will try some other way to solve this problem.

**Comment by ZX-lang** [comment]
> Exactly. I believe it is of great importance to the KD research.

When I use wrn40_2 as teacher and wrn40_1 as a student, I get a result that is not as good as the paper shows, could you please release the implementation of wrn

**Comment by DefangChen** [comment]
Note that the implementation of [wrn in CRD](https://github.com/HobbitLong/RepDistiller/blob/master/models/wrn.py) is somewhat problematic. It actually contains 38 layers rather than 40 layers. We have discussed this issue in line 18-24 of [resnet.py](https://github.com/DefangChen/SimKD/blob/main/models/resnet.py). (Due to the existence of this issue, I recommend to use resnet-depth x factor rather than wrn-depth x factor.)

In all our implementations, we actually use [resnet.py](https://github.com/DefangChen/SimKD/blob/main/models/resnet.py) to obtain resnet-38x2 using the [scripts](https://github.com/DefangChen/SimKD/blob/main/scripts/run_vanilla.sh) and rename it as wrn40_2 for consistency. (The pretrained resnet-38x2 model has been provided in GoogleDrive.) 

If you use a pre-trained teacher model with lower accuracy, it may lead to lower accuracy for all KD methods.

**Comment by ZX-lang** [comment]
> Note that the implementation of [wrn in CRD](https://github.com/HobbitLong/RepDistiller/blob/master/models/wrn.py) is somewhat problematic. It actually contains 38 layers rather than 40 layers. We have discussed this issue in line 18-24 of [resnet.py](https://github.com/DefangChen/SimKD/blob/main/models/resnet.py). (Due to the existence of this issue, I recommend to use resnet-depth x factor rather than wrn-depth x factor.)
> 
> In all our implementations, we actually use [resnet.py](https://github.com/DefangChen/SimKD/blob/main/models/resnet.py) to obtain resnet-38x2 using the [scripts](https://github.com/DefangChen/SimKD/blob/main/scripts/run_vanilla.sh) and rename it as wrn40_2 for consistency. (The pretrained resnet-38x2 model has been provided in GoogleDrive.)
> 
> If you use a pre-trained teacher model with lower accuracy, it may lead to lower accuracy for all KD methods.

Thank you. I will try it
